src/main/java/org/terasology/logic/health/HealthAuthoritySystem.java

```diff
@@ -0,0 +1,155 @@
+                continue;
+            }
+
+            int healAmount = 0;
+            healAmount = regenerateHealth(health, healAmount);
+
+            checkHealed(entity, health, healAmount);
+        }
+    }
+
+    private int regenerateHealth(HealthComponent health, int healAmount) {
+        int newHeal = healAmount;
+        while (time.getGameTimeInMs() >= health.nextRegenTick) {
+            newHeal++;
+            health.nextRegenTick = health.nextRegenTick + (long) (1000 / health.regenRate);
+        }
+        return newHeal;
+    }
+
+    private void checkHealed(EntityRef entity, HealthComponent health, int healAmount) {
+        if (healAmount > 0) {
+            checkHeal(entity, healAmount, entity);
+            entity.saveComponent(health);
+        }
+    }
+
+    private void checkHeal(EntityRef entity, int healAmount, EntityRef instigator) {
+        BeforeHealEvent beforeHeal = entity.send(new BeforeHealEvent(healAmount, instigator));
+        if (!beforeHeal.isConsumed()) {
+            int modifiedAmount = calculateTotal(beforeHeal.getBaseHeal(), beforeHeal.getMultipliers(), beforeHeal.getModifiers());
+            if (modifiedAmount > 0) {
+                doHeal(entity, modifiedAmount, instigator);
+            } else if (modifiedAmount < 0) {
+                // TODO: Add damage events
+                // doDamage(entity, -modifiedAmount, EngineDamageTypes.HEALING.get(), instigator, EntityRef.NULL);
+            }
+        }
+    }
+
+    private void doHeal(EntityRef entity, int healAmount, EntityRef instigator) {
+        HealthComponent health = entity.getComponent(HealthComponent.class);
+        if (health != null) {
+            int healedAmount = Math.min(health.currentHealth + healAmount, health.maxHealth) - health.currentHealth;
+            health.currentHealth += healedAmount;
+            entity.saveComponent(health);
+            entity.send(new OnHealedEvent(healAmount, healedAmount, instigator));
+            if (health.currentHealth == health.maxHealth) {
+                entity.send(new FullHealthEvent(instigator));
+            }
+        }
+    }
+
+    private int calculateTotal(int base, TFloatList multipliers, TIntList modifiers) {
+        // For now, add all modifiers and multiply by all multipliers. Negative modifiers cap to zero, but negative
+        // multipliers remain (so damage can be flipped to healing)
+
+        float total = base;
+        TIntIterator modifierIter = modifiers.iterator();
+        while (modifierIter.hasNext()) {
+            total += modifierIter.next();
+        }
+        total = Math.max(0, total);
+        if (total == 0) {
+            return 0;
+        }
+        TFloatIterator multiplierIter = multipliers.iterator();
+        while (multiplierIter.hasNext()) {
+            total *= multiplierIter.next();
+        }
+        return TeraMath.floorToInt(total);
+
+    }
+
+    /** Handles DoHeal event to increase health of entity */
```

Comment:
One-liner javadoc is good for simple fields, but methods should always use the expanded version including doc for parameters etc :-)